### PR TITLE
Add Parakeet-RNNT

### PR DESCRIPTION
**Pull request by eustlb**
# What does this PR do?

This PR adds parakeet RNNT architecture by refactoring the TDT one already merged: TDT is an extension of RNNT, and it should be implemented as such in the lib.

original checkpoints:
- [nvidia/parakeet-rnnt-0.6b](https://huggingface.co/nvidia/parakeet-rnnt-0.6b)
- [nvidia/parakeet-rnnt-1.1b](https://huggingface.co/nvidia/parakeet-rnnt-1.1b)

### CODE REVIEW

File: docs/source/en/model_doc/parakeet.md [hunk 1]
```diff
@@ -33,6 +33,11 @@ Parakeet models, [introduced by NVIDIA NeMo](https://developer.nvidia.com/blog/p
     - 1D convolution projection from encoder hidden size to vocabulary size (for optimal NeMo compatibility).
     - CTC loss computation for training.
     - Greedy CTC decoding for inference.
+- [**ParakeetForRNNT**](#parakeetforrnnt): a Fast Conformer Encoder + an RNN-T (RNN Transducer) decoder
```
Comment (vasqu) [inline docs/source/en/model_doc/parakeet.md:36]: Nit: any reason this is inbetween existing one?

File: docs/source/en/model_doc/parakeet.md [hunk 2]
```diff
@@ -86,6 +91,83 @@ print(processor.decode(outputs))
...
+Unlike TDT (which predicts a per-token duration), each RNN-T token is emitted at a single encoder frame, so its start and end span exactly one frame.
+
+```py
+from datasets import Audio, load_dataset
+from transformers import AutoModelForRNNT, AutoProcessor
+
+model_id = "nvidia/parakeet-rnnt-0.6b"
+revision = "refs/pr/4"
+processor = AutoProcessor.from_pretrained(model_id, revision=revision)
+model = AutoModelForRNNT.from_pretrained(model_id, revision=revision, dtype="auto", device_map="auto")
```
Comment (vasqu) [inline docs/source/en/model_doc/parakeet.md]: dtype auto is not needed (it's the default), no?

File: docs/source/en/model_doc/parakeet.md [hunk 3]
```diff
@@ -269,6 +351,38 @@ print("Loss:", outputs.loss.item())
...
+# passing `text` to the processor will prepare inputs' `labels` key
+inputs = processor(audio=speech_samples, text=text_samples, sampling_rate=processor.feature_extractor.sampling_rate)
+inputs.to(model.device, dtype=model.dtype)
+
+outputs = model(**inputs)
+print("Loss:", outputs.loss.item())
+outputs.loss.backward()
+```
+
+> [!NOTE]
```
Comment (vasqu) [inline docs/source/en/model_doc/parakeet.md]: I would put that before tbh

File: docs/source/en/model_doc/parakeet.md [hunk 4]
```diff
@@ -269,6 +351,38 @@ print("Loss:", outputs.loss.item())
...
+from datasets import Audio, load_dataset
+import torch
+from transformers import AutoModelForRNNT, AutoProcessor
+
+model_id = "nvidia/parakeet-rnnt-0.6b"
+revision = "refs/pr/4"
+NUM_SAMPLES = 4
+
+processor = AutoProcessor.from_pretrained(model_id, revision=revision)
+model = AutoModelForRNNT.from_pretrained(model_id, revision=revision, dtype=torch.bfloat16, device_map="auto")
```
Comment (vasqu) [inline docs/source/en/model_doc/parakeet.md]: Same re dtype is a bit weird but seeing the others already do that as well

File: src/transformers/loss/loss_rnnt.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+def rnnt_loss(
+    logits: torch.Tensor,
+    targets: torch.Tensor,
+    logit_lengths: torch.Tensor,
+    target_lengths: torch.Tensor,
+    blank_token_id: int,
+    reduction: str = "mean_volume",
+) -> torch.Tensor:
+    """
+    Compute standard RNN-T (RNN Transducer) loss (https://arxiv.org/abs/1211.3711).
```
Comment (vasqu) [inline src/transformers/loss/loss_rnnt.py]: no hf papers alternative?

File: src/transformers/loss/loss_rnnt.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        reduction: Loss reduction method. One of `"mean_volume"`, `"mean_batch"`, `"mean"`, `"sum"`, or `"none"`.
+
+    Returns:
+        Scalar loss tensor (or per-example losses if `reduction="none"`).
+
+    """
+
+    if not is_torchaudio_available():
+        raise ImportError(
+            "Computing the Parakeet RNN-T loss requires torchaudio. Install it with `pip install torchaudio`."
```
Comment (vasqu) [inline src/transformers/loss/loss_rnnt.py]:
```suggestion
            "Computing the RNN-T loss requires torchaudio. Install it with `pip install torchaudio`."
```

File: src/transformers/loss/loss_rnnt.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    blank_token_id: int,
+    reduction: str = "mean_volume",
+) -> torch.Tensor:
+    """
+    Compute standard RNN-T (RNN Transducer) loss (https://arxiv.org/abs/1211.3711).
+
+    Thin wrapper around [`torchaudio.functional.rnnt_loss`]. torchaudio is queried with `reduction="none"` to get
+    the per-sample negative log-likelihoods, and the requested reduction is applied here. The reduction names and
+    formulas mirror NeMo's `RNNTLoss` (the reference implementation used to train/finetune Parakeet), so that loss
+    magnitudes and gradient scaling match when finetuning a Parakeet checkpoint:
```
Comment (vasqu) [inline src/transformers/loss/loss_rnnt.py]:
```suggestion
    magnitudes and gradient scaling match when finetuning other RNNT models like Parakeet:
```

File: src/transformers/loss/loss_rnnt.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        )
+
+    target_lengths = target_lengths.to(logits.device)
+    losses = torchaudio.functional.rnnt_loss(
+        logits=logits.float().contiguous(),
+        targets=targets.to(logits.device).int(),
+        logit_lengths=logit_lengths.to(logits.device).int(),
+        target_lengths=target_lengths.int(),
+        blank=blank_token_id,
+        reduction="none",
```
Comment (vasqu) [inline src/transformers/loss/loss_rnnt.py:78]: Do we really need the contiguous call? At such big embeddings I expect this to be somewhat expensive

File: src/transformers/loss/loss_tdt.py [hunk 9]
```diff
@@ -29,7 +29,7 @@ def tdt_loss(
     blank_token_id: int,
     durations: list[int],
     sigma: float = 0.0,
-    reduction: str = "mean",
+    reduction: str = "mean_volume",
```
Comment (vasqu) [inline src/transformers/loss/loss_tdt.py]: Ig this is a bit breaking?

File: src/transformers/models/auto/configuration_auto.py [hunk 10]
```diff
@@ -37,6 +37,7 @@
     {
         "EvollaModel": "EvollaConfig",
         "mlcd": "MLCDVisionConfig",
+        "parakeet_tdt": "ParakeetTDTConfig",
```
Comment (vasqu) [inline src/transformers/models/auto/configuration_auto.py:40]: fixes from last time?

File: src/transformers/models/parakeet/configuration_parakeet.py [hunk 11]
```diff
@@ -187,4 +187,49 @@ def __post_init__(self, **kwargs):
         super().__post_init__(**kwargs)
 
 
-__all__ = ["ParakeetCTCConfig", "ParakeetEncoderConfig", "ParakeetTDTConfig"]
+@auto_docstring(checkpoint="nvidia/parakeet-tdt-0.6b-v3")
+@strict
+class ParakeetTDTConfig(ParakeetRNNTConfig):
```
Comment (vasqu) [inline src/transformers/models/parakeet/configuration_parakeet.py:188]: the git diff is not doing a good job here ngl but just checking if I understood this:
- RNNT does not use duration but otherwise mostly identical to tdt
- TDT hasnt changed but has loss reduction in the config added

The loss reduction is weird imo? Any reason we need this in the config and don't do it like in other models via kwargs?

File: src/transformers/models/parakeet/generation_parakeet.py [hunk 12]
```diff
@@ -20,7 +20,7 @@
 from ...utils import ModelOutput
 
 
-class ParakeetTDTDecoderCache:
+class ParakeetRNNTDecoderCache:
```
Comment (vasqu) [inline src/transformers/models/parakeet/generation_parakeet.py:23]: renames are breaking, any reason we really need this as its used for both either way

File: src/transformers/models/parakeet/generation_parakeet.py [hunk 13]
```diff
@@ -227,19 +236,51 @@ def prepare_inputs_for_generation(self, input_ids, *args, **kwargs):
         return model_inputs
 
     def generate(self, inputs=None, generation_config=None, **kwargs):
-        # TODO @eustlb: this is temporary — we're going to modularize generate to allow doing this cleanly.
```
Comment (vasqu) [inline src/transformers/models/parakeet/generation_parakeet.py:230]: not a todo anymore?

File: src/transformers/models/parakeet/modular_parakeet.py [hunk 14]
```diff
@@ -885,5 +881,55 @@ def forward(
             decoder_cache=decoder_cache,
         )
 
+    def loss_function(self, logits, labels, encoder_outputs, **kwargs):
```
Comment (vasqu) [inline src/transformers/models/parakeet/modular_parakeet.py]: this is weird, I would just add it to the forward

File: src/transformers/models/parakeet/modular_parakeet.py [hunk 15]
```diff
@@ -885,5 +881,55 @@ def forward(
...
+)
+class ParakeetForTDT(ParakeetTDTGenerationMixin, ParakeetForRNNT):
+    config: ParakeetTDTConfig
+
+    def __init__(self, config: ParakeetTDTConfig):
+        super().__init__(config)
+        self.joint = ParakeetTDTJointNetwork(config)
+        self.post_init()
+
+    def loss_function(self, logits, labels, encoder_outputs, **kwargs):
```
Comment (vasqu) [inline src/transformers/models/parakeet/modular_parakeet.py]: Ah it's a trick for modular? :D Yea still I'd rather have it explicit than to use "dirty" tricks

File: tests/models/parakeet/test_modeling_parakeet.py [hunk 16]
```diff
@@ -771,3 +778,332 @@ def test_tdt_model_integration_loss(self):
...
+            "feature-extraction": ParakeetEncoder,
+            "automatic-speech-recognition": ParakeetForRNNT,
+        }
+        if is_torch_available()
+        else {}
+    )
+
+    test_attention_outputs = False
+    test_resize_embeddings = False
+    test_torch_exportable = False
```
Comment (vasqu) [inline tests/models/parakeet/test_modeling_parakeet.py:958]: hmm new failure(?) not too important but would be nice to fix at some point

File: tests/models/parakeet/test_modeling_parakeet.py [hunk 17]
```diff
@@ -771,3 +778,332 @@ def test_tdt_model_integration_loss(self):
...
+        reason="ParakeetForRNNT has a custom generate() that is not fully compatible with GenerationTesterMixin"
+    )
+    def test_generation_tester_mixin_inheritance(self):
+        pass
+
+    @unittest.skip(reason="ParakeetForRNNT is a flat composite model without a separate base_model sub-module")
+    def test_model_base_model_prefix(self):
+        pass
+
+    @unittest.skip(reason="ParakeetForRNNT decoder is an LSTM prediction network without attention")
```
Comment (vasqu) [inline tests/models/parakeet/test_modeling_parakeet.py:1010]: Ig same as in tdt? across all skipped tests

File: tests/models/parakeet/test_modeling_parakeet.py
Comment (vasqu) [inline tests/models/parakeet/test_modeling_parakeet.py:1]: Do we need something like `TDTLossTest` for RNNT?

File: tests/models/parakeet/test_modeling_parakeet.py [hunk 18]
```diff
@@ -771,3 +778,332 @@ def test_tdt_model_integration_loss(self):
...
+        ds = self._dataset
+        speech_samples = ds.sort("id")[:num_samples]["audio"]
+        return [x["array"] for x in speech_samples]
+
+    @slow
+    def test_rnnt_model_integration(self):
+        """
+        reproducer: https://gist.github.com/eustlb/2b9a9a85ec447b176d14b23fc1484496#file-reproducer_single_rnnt-py
+        """
+        RESULTS_PATH = Path(__file__).parent.parent.parent / "fixtures/parakeet/expected_results_single_rnnt.json"
```
Comment (vasqu) [inline tests/models/parakeet/test_modeling_parakeet.py]: We can at least save the base path to the setup (`Path(__file__).parent.parent.parent`)

**Review by vasqu — APPROVED** [review]
Overall smaller comments, just some things to double check to make sure we are not breaking please. But great work otherwise 🫡

File: docs/source/en/model_doc/parakeet.md (hunk 1 above)
Comment (eustlb) [reply docs/source/en/model_doc/parakeet.md:36]: yep it is on purpose, TDT should be seen an extension of RNN-T

File: src/transformers/loss/loss_tdt.py (hunk 9 above)
Comment (eustlb) [reply src/transformers/loss/loss_tdt.py]: I really don't think this was used and this is aligned with NeMo codebase. But I am okay adressing this to a dedicated PR to have this clearer for users

File: docs/source/en/model_doc/parakeet.md (hunk 2 above)
Comment (eustlb) [reply docs/source/en/model_doc/parakeet.md]: yep indeed. I've also checked that from NeMo lib:
- ctc: bfloat16 (effectively in config)
- rnnt/ tdt : float32 (effectively in config)

so I am removing dtypes from the doc

File: src/transformers/loss/loss_rnnt.py (hunk 5 above)
Comment (eustlb) [reply src/transformers/loss/loss_rnnt.py]: indeed, thanks for catching!

File: src/transformers/models/parakeet/generation_parakeet.py (hunk 13 above)
Comment (eustlb) [reply src/transformers/models/parakeet/generation_parakeet.py:230]: definitely still!

File: src/transformers/models/parakeet/generation_parakeet.py (hunk 12 above)
Comment (eustlb) [reply src/transformers/models/parakeet/generation_parakeet.py:23]: I don't think it's been used at all, but still added
```python
class ParakeetTDTDecoderCache(ParakeetRNNTDecoderCache): ...
```

File: tests/models/parakeet/test_modeling_parakeet.py (hunk 16 above)
Comment (eustlb) [reply tests/models/parakeet/test_modeling_parakeet.py:958]: nope, just mirrored tdt here

File: tests/models/parakeet/test_modeling_parakeet.py (hunk 17 above)
Comment (eustlb) [reply tests/models/parakeet/test_modeling_parakeet.py:1010]: yep

File: tests/models/parakeet/test_modeling_parakeet.py (hunk 18 above)
Comment (eustlb) [reply tests/models/parakeet/test_modeling_parakeet.py]: yep, added [FIXTURES_DIR](https://github.com/huggingface/transformers/pull/46331/commits/965c04b3406c9dde2e3f733e3085c4a8b12454e8)

File: src/transformers/models/parakeet/configuration_parakeet.py (hunk 11 above)
Comment (eustlb) [reply src/transformers/models/parakeet/configuration_parakeet.py:188]: removed it and now to be set via kwargs 👌

File: src/transformers/models/parakeet/modular_parakeet.py (hunk 14 above)
Comment (eustlb) [reply src/transformers/models/parakeet/modular_parakeet.py]: Agree this was a bit `circumvoluted`, thanks for pushing back! This strange logic was not due to modular but rather avoid duplicating lines via the inheritance, but here re-writing the full forward is really the cleanest. That's what I went to in the end

File: tests/models/parakeet/test_modeling_parakeet.py
Comment (eustlb) [reply tests/models/parakeet/test_modeling_parakeet.py:1]: added !

File: src/transformers/models/parakeet/modular_parakeet.py (hunk 15 above)
Comment (eustlb) [reply src/transformers/models/parakeet/modular_parakeet.py]: cf above comment, went for a cleaner approach :) thanks!

File: src/transformers/loss/loss_rnnt.py (hunk 8 above)
Comment (eustlb) [reply src/transformers/loss/loss_rnnt.py:78]: torchaudio's rejects non-contiguous :/

File: src/transformers/models/parakeet/generation_parakeet.py (hunk 12 above)
Comment (vasqu) [reply src/transformers/models/parakeet/generation_parakeet.py:23]: Just better safe than sorry

**Review by vasqu — APPROVED** [review]
Looking good! Thanks for addressing and nice work

Let's use run-slow but I think we are good to merge afterwards 🫡

File: src/transformers/models/auto/configuration_auto.py (hunk 10 above)
Comment (eustlb) [reply src/transformers/models/auto/configuration_auto.py:40]: more that now `ParakeetTDTConfig subclasses` ParakeetRNNTConfig, and since AST does  non-recursive base.id == "PreTrainedConfig", it's not discovered so that's necessary

**Comment by eustlb** [comment]
run-slow: parakeet

**Comment by vasqu** [comment]
https://app.circleci.com/pipelines/github/huggingface/transformers/177680/workflows/3c0b8544-7b23-49a9-ae21-0476dd5961a5/jobs/2350759 might be a real regression but I can merge afterwards, just ping me. CI is super flaky :cry:

**Comment by eustlb** [comment]
It was indeed! Now fixed, merging 🎉 thank you so much for the reviews @vasqu !!!! 🫶
